File: research/ignition-advance-static/model/engine.py
"""Параметры двигателя из входного JSON."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class EngineParams:
    bore_mm: float
    stroke_mm: float
    rod_length_mm: float | None
    cylinder_count: int
    displacement_cc: float | None
    compression_ratio: float

    valves_per_cylinder: int
    spark_location: str
    chamber_type: str

    intake_duration_deg: float | None
    exhaust_duration_deg: float | None
    overlap_deg: float | None

    fuel: str
    aspiration: str

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> EngineParams:
        return cls(
            bore_mm=float(data["bore_mm"]),
            stroke_mm=float(data["stroke_mm"]),
            rod_length_mm=_optional_float(data.get("rod_length_mm")),
            cylinder_count=int(data["cylinder_count"]),
            displacement_cc=_optional_float(data.get("displacement_cc")),
            compression_ratio=float(data["compression_ratio"]),
            valves_per_cylinder=int(data["valves_per_cylinder"]),
            spark_location=str(data["spark_location"]),
            chamber_type=str(data["chamber_type"]),
            intake_duration_deg=_optional_float(data.get("intake_duration_deg")),
            exhaust_duration_deg=_optional_float(data.get("exhaust_duration_deg")),
            overlap_deg=_optional_float(data.get("overlap_deg")),
            fuel=str(data.get("fuel", "gasoline_95")),
            aspiration=str(data.get("aspiration", "naturally_aspirated")),
        )

    @property
    def stroke_ratio(self) -> float:
        return self.stroke_mm / self.bore_mm

    @property
    def is_forced_induction(self) -> bool:
        return self.aspiration in ("turbocharged", "supercharged")


def _optional_float(value: Any) -> float | None:
    if value is None:
        return None
    return float(value)

File: crates/rusefui-runtime/src/ignition_map/engine.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EngineParams {
    pub bore_mm: f64,
    pub stroke_mm: f64,
    #[serde(default)]
    pub rod_length_mm: Option<f64>,
    #[serde(default = "default_cylinder_count")]
    pub cylinder_count: u32,
    #[serde(default)]
    pub displacement_cc: Option<f64>,
    pub compression_ratio: f64,
    #[serde(default = "default_valves")]
    pub valves_per_cylinder: u32,
    #[serde(default = "default_spark_location")]
    pub spark_location: String,
    #[serde(default = "default_chamber_type")]
    pub chamber_type: String,
    #[serde(default)]
    pub intake_duration_deg: Option<f64>,
    #[serde(default)]
    pub exhaust_duration_deg: Option<f64>,
    #[serde(default)]
    pub overlap_deg: Option<f64>,
    #[serde(default = "default_fuel")]
    pub fuel: String,
    #[serde(default = "default_aspiration")]
    pub aspiration: String,
}

fn default_cylinder_count() -> u32 {
    4
}

fn default_valves() -> u32 {
    4
}

fn default_spark_location() -> String {
    "center".into()
}

fn default_chamber_type() -> String {
    "pentroof".into()
}

fn default_fuel() -> String {
    "gasoline_95".into()
}

fn default_aspiration() -> String {
    "naturally_aspirated".into()
}

impl Default for EngineParams {
    fn default() -> Self {
        Self {
            bore_mm: 86.0,
            stroke_mm: 86.0,
            rod_length_mm: None,
            cylinder_count: 4,
            displacement_cc: None,
            compression_ratio: 10.0,
            valves_per_cylinder: 4,
            spark_location: default_spark_location(),
            chamber_type: default_chamber_type(),
            intake_duration_deg: None,
            exhaust_duration_deg: None,
            overlap_deg: None,
            fuel: default_fuel(),
            aspiration: default_aspiration(),
        }
    }
}
